transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,556 @@
+                        );
+                        return ControlFlow::Break(());
+                    }
+                }
+
+                let event_back = self.endpoint.handle_event(connection_id, event);
+
+                if let Some(event_back) = event_back {
+                    debug_assert!(!is_drained_event);
+                    if let Some(sender) = self.alive_connections.get_mut(&connection_id) {
+                        // We clone the sender to guarantee that there will be at least one
+                        // free slot to send the event.
+                        // The channel can not grow out of bound because an `event_back` is
+                        // only sent if we prior received an event from the same connection.
+                        // If the connection is busy, it won't sent us any more events to handle.
+                        let _ = sender.clone().start_send(event_back);
+                    } else {
+                        log::error!("State mismatch: event for closed connection");
+                    }
+                }
+            }
+
+            // Data needs to be sent on the UDP socket.
+            ToEndpoint::SendUdpPacket(transmit) => self.next_packet_out = Some(transmit),
+            ToEndpoint::Decoupled => self.handle_decoupling()?,
+        }
+        ControlFlow::Continue(())
+    }
+
+    /// Handle datagram received on the socket.
+    /// The datagram content was written into the `socket_recv_buffer`.
+    fn handle_datagram(&mut self, packet: BytesMut, packet_src: SocketAddr) -> ControlFlow<()> {
+        let local_ip = self.channel.socket_addr.ip();
+        // TODO: ECN bits aren't handled
+        let (connec_id, event) =
+            match self
+                .endpoint
+                .handle(Instant::now(), packet_src, Some(local_ip), None, packet)
+            {
+                Some(event) => event,
+                None => return ControlFlow::Continue(()),
+            };
+        match event {
+            quinn_proto::DatagramEvent::ConnectionEvent(event) => {
+                // Redirect the datagram to its connection.
+                if let Some(sender) = self.alive_connections.get_mut(&connec_id) {
+                    match sender.try_send(event) {
+                        Ok(()) => {}
+                        Err(err) if err.is_disconnected() => {
+                            // Connection was dropped by the user.
+                            // Inform the endpoint that this connection is drained.
+                            self.endpoint
+                                .handle_event(connec_id, quinn_proto::EndpointEvent::drained());
+                            self.alive_connections.remove(&connec_id);
+                        }
+                        Err(err) if err.is_full() => {
+                            // Connection is too busy. Drop the datagram to back-pressure the remote.
```

Comment:
The comment itself is outdated,  but still a valid discussion.
The concern is that after handling an udp packet, the endpoint returns here an `ConnectionEvent` for a known connection. This type has two variants:
1. `Datagram`: a forwarded datagram -> safe to drop
2. `NewIdentifiers`: new connection identifiers -> not safe to drop!

By type, the returned event may theoretically by both variants, which is why the original comment said it was not safe to drop. However, in practice only the `Datagram` variant is returned here [^1]. This is also stated in the docs on `DatagramEvent::ConnectionEvent`: " /// The datagram is redirected to its `Connection`". While it is not ideal to rely on this I am hesitant to add more complex handling for a case that can never happen in practice. 

[^1]: https://github.com/quinn-rs/quinn/blob/cd2cf756acc9e7774d2c40992e7a355d0fb51b74/quinn-proto/src/endpoint.rs#L234-L245

Reply:
Dang it's a shame the variants are `pub` because at least we can debug assert our understanding of the code.

My worry now isn't about the behavior of quinn is today, but the behavior tomorrow. I'm afraid that this will change and we won't notice because the type didn't change, and then we'll drop critical events and be in inconsistent states that are hard to debug.

I'm not saying we should handle the `NewIdentifiers` case, but is it possible to add a check to make sure we're only dropping datagrams (even if only enabled in debug builds)

Reply:
I completely agree. I am also not happy about how it is write now, but the API currently does not give us any information about the event. Added a debug message for now, but to properly solve this long-term we could propose an method upstream `ConnectionEvent::is_datagram_event`, and `debug_assert!` that then. Will write it down to the tracking issue.

Reply:
It is a bit hacky but you could `Debug` print the event and assert that it is the Datagram variant :)

Reply:
If you do it all within the `debug_assert!` macro, it won't be compiled in for production so it is zero-cost there :)

Reply:
Definitely hacky :D but I am not able to come up with a better solution and it's still better than nothing. Added in 445a2a4102f38b80916d0d5667305668d282e419.